Go tutorial 🐿️ (#8904)
Unleash/unleash: website/docs/feature-flag-tutorials/golang/golang-examples.md

---
title: Go feature flag examples
slug: /feature-flag-tutorials/golang/examples
---

> In our [Go feature flag tutorial](/feature-flag-tutorials/go), we implemented a simple feature flag that could be turned on and off. This document will walk you through other examples of what can be achieved using feature flags in Go.

Author: That's exactly it, this was mostly for reference purposes. So that people have easy blocks of code to copy-paste for various actions in their favourite languages. We've done this for every language we support so far.

We built many features into Unleash, our open-source feature flag platform, to address the complexities of releasing code. This tutorial will explore the following:

## Gradual Rollouts for Go Apps

It is common to use feature flags to roll out changes to a percentage of users, and we can use Unleash to do that too.

Here's how you do it using our API:

```go
payload := map[string]interface{}{
    "name":     "flexibleRollout",
    "disabled": false,
    "constraints": []interface{}{},
    "variants":    []interface{}{},
    "parameters": map[string]interface{}{
        "groupId":    "delete_survey_flag",
        "rollout":    "50",
        "stickiness": "sessionId",
    },
    "segments": []interface{}{},
}

jsonData, err := json.Marshal(payload)

url := fmt.Sprintf("%s/api/admin/projects/%s/features/%s/environments/%s/strategies",
    unleashURL, projectID, featureName, environment)

req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))

req.Header.Set("Content-Type", "application/json")
req.Header.Set("Authorization", apiKey)

client := &http.Client{}
resp, err := client.Do(req)
defer resp.Body.Close()

> ```

Author: It was there for historical reason, if it's not something typical then let's kill it 🔥

---
title: Go Feature Flag Examples
slug: /feature-flag-tutorials/golang/examples
---

In our [Go feature flag tutorial](/feature-flag-tutorials/golang), we implemented a simple feature flag that could be turned on and off. This document will walk you through some common examples of using feature flags in Go.

We built many features into Unleash, our open-source feature flag platform, to address the complexities of releasing code. This tutorial will explore the following:

> - [Gradual Rollouts for Go Apps](#gradual-rollouts-for-go-apps)

Reviewer: Let's use title case and drop the toc. You could still do a summary of roughly what is covered in the tutorial, but I don't think we don't need to list and link to each section

- [Canary Deployments in Go](#canary-deployments-in-go)
  - [What is a canary deployment?](#what-is-a-canary-deployment)
  - [Why use canary deployments?](#why-use-canary-deployments)
  - [How to do canary deployments with a feature flag in Go?](#how-to-do-canary-deployments-with-a-feature-flag-in-go)
  - [Configure strategy constraints for canary deployments](#configure-strategy-constraints-for-canary-deployments)
- [Server-side A/B Testing in Go](#server-side-ab-testing-in-go)
- [Feature Flag Analytics and Reporting in Go](#feature-flag-analytics-and-reporting-in-go)
  - [Enable impression data events in Go](#enable-impression-data-events-in-go)
- [Application Metrics and Monitoring for Go apps](#application-metrics-and-monitoring-for-go-apps)
- [Feature Flag Audit Logs in Go](#feature-flag-audit-logs-in-go)
- [Flag Automation and Workflow Integration for Go Apps](#flag-automation-and-workflow-integration-for-go-apps)

## Gradual Rollouts for Go Apps

It is common to use feature flags to roll out changes to a percentage of users, and we can use Unleash to do that too.

> Navigate to the gradual rollout page where you can edit your strategy.

Reviewer: What do you mean by 'gradual rollout page'?

```suggestion
Go to the gradual rollout page where you can edit your strategy.
```

![The "edit strategy" button uses a pencil icon and is located on every strategy.](/img/react-ex-grad-rollout-edit.png)

> Adjust the percentage of users to 50% or whichever value you choose, and refresh your app in the browser to see if your user has the new feature experience. This might take 30 seconds for a server-side app to propagate.

Reviewer: I don't get this 🤔 why would a 50% rollout guarantee that my user has the new experience?

### Why use canary deployments?

Canary deployments are a safer and more gradual way to make changes in software development. They help find any abnormalities and align with the agile process for faster releases and quick reversions.

### How to do canary deployments with a feature flag in Go?

Feature flags provide the same benefits as canary deployments but with more granular control:

-   Precisely target specific user segments for feature rollouts.
-   Maintain session consistency (stickiness) if needed.
-   Test multiple features independently on different user groups simultaneously.
-   With feature flags, you can separate feature releases from deployments.

Often, canary deployments are managed at the load balancer level while feature flags act at the application level. In some instances, rolling out groups of features together behind a feature flag can serve the purpose of a canary deployment.

### Configure strategy constraints for canary deployments

> Let's update our existing gradual rollout strategy using Go to add environment-based constraints:

Reviewer: We could link to constraints docs

```go
payload := map[string]interface{}{
    "name":     "flexibleRollout",
    "disabled": false,
    "constraints": []map[string]interface{}{
        {
            "values":         []string{"production"},
            "inverted":       false,
            "operator":       "NOT_IN",
            "contextName":    "environment",
            "caseInsensitive": false,
        },
    },
    "variants": []interface{}{},
    "parameters": map[string]interface{}{
        "groupId":    "delete_survey_flag",
        "rollout":    "50",
        "stickiness": "sessionId",
    },
    "segments": []interface{}{},
}

jsonData, err := json.Marshal(payload)

url := fmt.Sprintf("%s/api/admin/projects/%s/features/%s/environments/%s/strategies/%s",
    unleashURL, projectID, featureName, environment, strategyID)

req, err := http.NewRequest("PUT", url, bytes.NewBuffer(jsonData))

req.Header.Set("Content-Type", "application/json")
req.Header.Set("Authorization", apiKey)

client := &http.Client{}
resp, err := client.Do(req)
defer resp.Body.Close()
```

## Server-side A/B Testing in Go

> A/B testing allows teams to test multiple versions of a feature with different user groups. In Go, we can implement this using strategy variants. Here's how to update a strategy to enable A/B testing for 50% of users:

Reviewer: ```suggestion
A/B testing allows you to test multiple versions of a feature with different user groups. In Go, we can implement this using strategy variants. Here's how to update a strategy to enable A/B testing for 50% of users:
```

Reviewer: We could link to variants docs

```go
payload := map[string]interface{}{
    "name":  "flexibleRollout",
    "title": "",
    "constraints": []interface{}{},
    "parameters": map[string]interface{}{
        "rollout":    "50",
        "stickiness": "default",
        "groupId":    "",
    },
    "variants":  []interface{}{},
    "segments":  []interface{}{},
    "disabled": false,
}

jsonData, err := json.Marshal(payload)

url := fmt.Sprintf("%s/api/admin/projects/%s/features/%s/environments/%s/strategies/%s",
    unleashURL, projectID, featureName, environment, strategyID)

req, err := http.NewRequest("PUT", url, bytes.NewBuffer(jsonData))

req.Header.Set("Content-Type", "application/json")
req.Header.Set("Authorization", apiKey)

client := &http.Client{}
resp, err := client.Do(req)
defer resp.Body.Close()
```

## Feature Flag Analytics and Reporting in Go

> Shipping code is one thing, but monitoring your applications is another aspect of managing code to account for. Some things to monitor could be:

Reviewer: ```suggestion
Shipping code is one thing, but monitoring your applications is another aspect of managing code. For example, you could use feature flag analytics to monitor performance metrics or track user behavior.
```

> -   Performance metrics

Reviewer: ```suggestion
```

> -   Tracking user behavior

Reviewer: ```suggestion
```

> ### Enable impression data events in Go

Reviewer: The previous section could use a sentence introducing impression data.

        return fmt.Errorf("error decoding response: %v", err)
    }

    // Process metrics data
    fmt.Printf("Metrics for %s: %+v\n", featureName, metricsData)
    return nil
}
```

## Feature Flag Audit Logs in Go

Implementing audit logging in your Go application helps track changes to feature flags over time. Here's how to retrieve and process audit logs:

```go
func getAuditLogs(unleashURL, featureName, apiKey string) error {
    url := fmt.Sprintf("%s/api/admin/events/%s", unleashURL, featureName)

    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return fmt.Errorf("error creating request: %v", err)
    }

    req.Header.Add("Accept", "application/json")
    req.Header.Add("Authorization", apiKey)

    client := &http.Client{}
    resp, err := client.Do(req)
    if err != nil {
        return fmt.Errorf("error making request: %v", err)
    }
    defer resp.Body.Close()

    body, err := io.ReadAll(resp.Body)
    if err != nil {
        return fmt.Errorf("error reading response: %v", err)
    }

    fmt.Println(string(body))
    return nil
}
```

## Flag Automation and Workflow Integration for Go Apps

> An advanced use case for leveraging feature flags at scale is automating them as part of your development workflow.

Reviewer: ```suggestion
An advanced use case for using feature flags at scale is automating them as part of your development workflow.
```
